### PR TITLE
Use v:echospace to avoid hit-Enter prompts

### PR DESCRIPTION
**Pull request by blueyed**
Fixes https://github.com/tpope/vim-dispatch/issues/166
Followup-to https://github.com/tpope/vim-dispatch/pull/167

### CODE REVIEW

File: autoload/dispatch.vim [hunk 1]
```diff
@@ -363,6 +363,54 @@ function! s:postfix(request) abort
...
+    if &showcmd
+      let max_cmd_len -= 10
+      if !&ruler || last_has_status
+        let max_cmd_len -= 1
+      endif
+    endif
+  endif
+
+  let msg_len = len(substitute(a:msg, '.', '.', 'g'))
+  if msg_len > max_cmd_len
```
Comment (blueyed) [inline autoload/dispatch.vim]: Could also use `strwidth` maybe if available?

File: autoload/dispatch.vim [hunk 2]
```diff
@@ -363,6 +363,54 @@ function! s:postfix(request) abort
...
 
+function! s:echo_without_hitenter(msg, suffix) abort
+  if exists('v:echospace')
+    if strwidth(a:msg.' '.a:suffix) > v:echospace
+      let msg = printf('%.*S...', v:echospace - 3 - len(a:suffix), a:msg)
+    else
+      let msg = a:msg
+    endif
+    echom msg.' '.a:suffix
+    return
```
Comment (blueyed) [inline autoload/dispatch.vim]: Could `echo`/`echom` directly (also below).

File: autoload/dispatch.vim [hunk 3]
```diff
@@ -363,6 +363,54 @@ function! s:postfix(request) abort
   return '(' . a:request.handler.'/'.(!empty(pid) ? pid : '?') . ')'
 endfunction
 
+function! s:echo_without_hitenter(msg, suffix) abort
```
Comment (tpope) [inline autoload/dispatch.vim]: The message is "press enter", not "hit enter", but either way it's too much of a mouthful. Try `echo_truncated()`

File: autoload/dispatch.vim [hunk 4]
```diff
@@ -363,6 +363,54 @@ function! s:postfix(request) abort
...
+    if strwidth(a:msg.' '.a:suffix) > v:echospace
+      let msg = printf('%.*S...', v:echospace - 3 - len(a:suffix), a:msg)
+    else
+      let msg = a:msg
+    endif
+    echom msg.' '.a:suffix
+    return
+  endif
+
+  redraw
```
Comment (tpope) [inline autoload/dispatch.vim]: Why does the old logic require a redraw but the new logic doesn't?

File: autoload/dispatch.vim (hunk 1 above)
Comment (tpope) [reply autoload/dispatch.vim]: If it ain't broke don't fix it.

File: autoload/dispatch.vim (hunk 4 above)
Comment (blueyed) [reply autoload/dispatch.vim]: Should be done with both probably.

File: autoload/dispatch.vim [hunk 5]
```diff
@@ -363,6 +363,55 @@ function! s:postfix(request) abort
   return '(' . a:request.handler.'/'.(!empty(pid) ? pid : '?') . ')'
 endfunction
 
+function! s:echo_truncated(msg, suffix) abort
+  redraw
+
+  if exists('v:echospace')
+    if strwidth(a:msg.' '.a:suffix) > v:echospace
+      let msg = printf('%.*S...', v:echospace - 3 - len(a:suffix), a:msg)
```
Comment (tpope) [inline autoload/dispatch.vim]: This should use the `<` notation from the other format, not an ellipsis.

**Comment by petobens** [comment]
Any chance of moving this forward @blueyed @tpope ? It would be a really nice addition... Thanks!

File: autoload/dispatch.vim (hunk 1 above)
Comment (axelf4) [reply autoload/dispatch.vim]: Why is the `substitute()` needed here?

**Comment by blueyed** [comment]
@petobens 
I feel like if I'd fix/amend it @tpope would do it himself anyway - so it's waiting on him.

**Comment by petobens** [comment]
OK. Let's wait for @tpope then. Thanks

**Comment by tpope** [comment]
If I was going to do it myself I would have done it myself rather than explicitly requesting changes.

File: autoload/dispatch.vim (hunk 1 above)
Comment (tpope) [reply autoload/dispatch.vim]: Character length rather than byte length.

**Comment by petobens** [comment]
Ok @blueyed can you wrap this up then? It's really annoying to press enter. Thank you so much in advance

**Comment by telemachus** [comment]
@blueyed I merged this PR locally to see if it fixed the problem, and at least here it actually made it worse. I.e., with the earlier version I had the `Press ENTER` problem with some very long-named files, but now I have it with all files.

Perhaps it has something to do with my settings though? (`shortmess=atIF` and `cmdheight=1`, in case that helps.)

**Comment by lourenci** [comment]
Hi, @blueyed. Is there some advance on this? Thank you for this.

**Pull request closed without merging**